<a href="https://colab.research.google.com/github/poojajangid19/Data-Analyst/blob/main/Pandas_3_with_Weather_Data_and_Uber_Case_Study.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Content

  - **Restructuring data**
    - `pd.melt()`
    - `pd.pivot_table()`


  - **Writing to a file**


  - **Basic EDA (Uber usecase)**:
    
    - Importance of EDA
    
    - Basic info of data:
      - `df.shape`
      - `df.info`
      - `df.head`
    
    - Handling datetime

In [12]:
!gdown 1wG8GaMi1hhfY6DsUPFjFtDYkumA-m2Xj

Downloading...
From: https://drive.google.com/uc?id=1wG8GaMi1hhfY6DsUPFjFtDYkumA-m2Xj
To: /content/weather.csv
100% 12.0k/12.0k [00:00<00:00, 38.5MB/s]


In [13]:
import pandas as pd

weather = pd.read_csv('weather.csv')
weather.head()

year  month element       day1       day2       day3       day4       day5  \
0  2018      1     max  17.573016  19.796815  22.412495  17.813163  20.165825   
1  2018      1     min  22.725760  21.007865  17.730792  18.045290  20.766734   
2  2018      2     max  19.015120  19.261805  17.510713  21.080425  17.915749   
3  2018      2     min  18.653843  22.818600  21.842673  21.958159  22.523078   
4  2018      3     max  20.741115  19.704016  17.039811  20.703908  22.714125   

        day6       day7  ...      day22      day23      day24      day25  \
0  17.060539  22.736134  ...  22.629226  19.013674  17.831558  17.018316   
1  18.656651  22.607481  ...  21.163491  17.371653  17.194530  18.964190   
2  19.082145  18.056023  ...  21.254968  20.346324  20.036019  22.926773   
3  18.535469  19.636158  ...  19.711974  20.586232  18.780840  20.065633   
4  17.205000  19.079503  ...  19.528963  19.563689  22.658997  22.446357   

       day26      day27      day28      day29      day30      day31  
0  21.837387  19.700255  19.084056  18.534690  20.787275  21.117423  
1  17.892385  17.646139  19.844125  22.729859  19.768917  20.913443  
2  17.533172  20.308927  20.438899  17.005806  18.346534  20.878165  
3  18.293548  20.823661  18.056112  22.748292  22.470172  18.602007  
4  17.911240  22.591993  17.406800  19.118712  22.767154  17.119452  

[5 rows x 34 columns]

In [14]:
weather.shape

(22, 34)

#### Takeaway:

- **Usually** no. of rows are more than no. of columns; But in this case, **no. of columns > no. of rows**

- The **columns are `day1`, `day2`, `day3`, ... so on**

- `max` and `min` **value of each day** of the month is **in a separate row**



#### Restructuring DataFrame to columns [`year`, `month`, `day`, `min`, `max`]
- `pd.melt()`




In [15]:
weather_melt = pd.melt(weather,
                       id_vars=['year', 'month', 'element'],
                       var_name="day",
                       value_name="temp")

# We'll store the result in a new variable

weather_melt

year  month element    day       temp
0    2018      1     max   day1  17.573016
1    2018      1     min   day1  22.725760
2    2018      2     max   day1  19.015120
3    2018      2     min   day1  18.653843
4    2018      3     max   day1  20.741115
..    ...    ...     ...    ...        ...
677  2018     10     min  day31  21.691537
678  2018     11     max  day31  20.750438
679  2018     11     min  day31  18.939767
680  2018     12     max  day31  19.648924
681  2018     12     min  day31  18.775539

[682 rows x 5 columns]


#### Notice:

- It's **using 2 rows for a single day**
    - One for **`min temp`** and one for **`max temp`**
    
    



####  We want to **split one single column into multiple columns**

`year | month | day | min_temp | max_temp`



#### - `pivot_table()`

Refer: https://pandas.pydata.org/docs/reference/api/pandas.pivot_table.html

- Pass in the **column names that we DON'T want to change**


- Pass in the **column name that we want to split**
    - **Labels of new columns will come from the column we want to split**


- Pass in the **column from which we will get our values**

In [16]:
weather_tidy = weather_melt.pivot(index=['year', 'month', 'day'],
                                        columns = 'element',
                                        values='temp')

weather_tidy

element                 max        min
year month day                        
2018 1     day1   17.573016  22.725760
           day10  19.067288  19.931129
           day11  19.361002  22.598325
           day12  20.982134  17.715137
           day13  21.668005  17.940334
...                     ...        ...
     12    day5   21.375349  20.865535
           day6   17.992885  20.310116
           day7   19.683359  20.531823
           day8   20.477046  19.310346
           day9   20.210640  22.820992

[341 rows x 2 columns]

**Takeaway:**
- Notice that a multi-index df has been created.
- We can change this using **`reset_index()`**

In [17]:
weather_tidy = weather_tidy.reset_index()
weather_tidy

element  year  month    day        max        min
0        2018      1   day1  17.573016  22.725760
1        2018      1  day10  19.067288  19.931129
2        2018      1  day11  19.361002  22.598325
3        2018      1  day12  20.982134  17.715137
4        2018      1  day13  21.668005  17.940334
..        ...    ...    ...        ...        ...
336      2018     12   day5  21.375349  20.865535
337      2018     12   day6  17.992885  20.310116
338      2018     12   day7  19.683359  20.531823
339      2018     12   day8  20.477046  19.310346
340      2018     12   day9  20.210640  22.820992

[341 rows x 5 columns]

## Writing to file



In [18]:
weather_tidy.to_csv('weather_tidy.csv', sep=",")
weather_tidy

element  year  month    day        max        min
0        2018      1   day1  17.573016  22.725760
1        2018      1  day10  19.067288  19.931129
2        2018      1  day11  19.361002  22.598325
3        2018      1  day12  20.982134  17.715137
4        2018      1  day13  21.668005  17.940334
..        ...    ...    ...        ...        ...
336      2018     12   day5  21.375349  20.865535
337      2018     12   day6  17.992885  20.310116
338      2018     12   day7  19.683359  20.531823
339      2018     12   day8  20.477046  19.310346
340      2018     12   day9  20.210640  22.820992

[341 rows x 5 columns]

## Exploratory Data Analysis - Uber Drives Data


  - It include performing operations on data like:

    - **Cleaning** the data

    - Dealing with **missing values**

    - **Adding** a few **columns** that might be helpful

    - **Restructuring** the data etc..




In [19]:
!gdown 1TL2hWkMWtD1ExVgaQhWP6A2swR8F8cVB

Downloading...
From: https://drive.google.com/uc?id=1TL2hWkMWtD1ExVgaQhWP6A2swR8F8cVB
To: /content/UberDrives.csv
100% 86.4k/86.4k [00:00<00:00, 85.4MB/s]


In [20]:
data = pd.read_csv('UberDrives.csv')
data

START_DATE*         END_DATE* CATEGORY*            START*  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
...                ...               ...       ...               ...   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                 STOP*   MILES*         PURPOSE*  
0          Fort Pierce      5.1   Meal/Entertain  
1          Fort Pierce      5.0              NaN  
2          Fort Pierce      4.8  Errand/Supplies  
3          Fort Pierce      4.7          Meeting  
4      West Palm Beach     63.7   Customer Visit  
...                ...      ...              ...  
1151  Unknown Location      3.9   Temporary Site  
1152  Unknown Location     16.2          Meeting  
1153           Gampaha      6.4   Temporary Site  
1154         Ilukwatta     48.2   Temporary Site  
1155               NaN  12204.7              NaN  

[1156 rows x 7 columns]

In [21]:
data.shape

(1156, 7)

In [22]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1156 entries, 0 to 1155
Data columns (total 7 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   START_DATE*  1156 non-null   object 
 1   END_DATE*    1155 non-null   object 
 2   CATEGORY*    1155 non-null   object 
 3   START*       1155 non-null   object 
 4   STOP*        1155 non-null   object 
 5   MILES*       1156 non-null   float64
 6   PURPOSE*     653 non-null    object 
dtypes: float64(1), object(6)
memory usage: 63.3+ KB


In [23]:
data.describe()

MILES*
count   1156.000000
mean      21.115398
std      359.299007
min        0.500000
25%        2.900000
50%        6.000000
75%       10.400000
max    12204.700000

In [24]:
data.head()

START_DATE*       END_DATE* CATEGORY*       START*            STOP*  \
0  1/1/2016 21:11  1/1/2016 21:17  Business  Fort Pierce      Fort Pierce   
1   1/2/2016 1:25   1/2/2016 1:37  Business  Fort Pierce      Fort Pierce   
2  1/2/2016 20:25  1/2/2016 20:38  Business  Fort Pierce      Fort Pierce   
3  1/5/2016 17:31  1/5/2016 17:45  Business  Fort Pierce      Fort Pierce   
4  1/6/2016 14:42  1/6/2016 15:49  Business  Fort Pierce  West Palm Beach   

   MILES*         PURPOSE*  
0     5.1   Meal/Entertain  
1     5.0              NaN  
2     4.8  Errand/Supplies  
3     4.7          Meeting  
4    63.7   Customer Visit

In [25]:
data.tail()

START_DATE*         END_DATE* CATEGORY*            START*  \
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                 STOP*   MILES*        PURPOSE*  
1151  Unknown Location      3.9  Temporary Site  
1152  Unknown Location     16.2         Meeting  
1153           Gampaha      6.4  Temporary Site  
1154         Ilukwatta     48.2  Temporary Site  
1155               NaN  12204.7             NaN

#### Is there any column having missing values?



In [26]:
data.isnull().sum()

START_DATE*      0
END_DATE*        1
CATEGORY*        1
START*           1
STOP*            1
MILES*           0
PURPOSE*       503
dtype: int64

#### Are there any duplicates values?

In [27]:
data[data.duplicated()]

START_DATE*        END_DATE* CATEGORY*  START* STOP*  MILES* PURPOSE*
492  6/28/2016 23:34  6/28/2016 23:59  Business  Durham  Cary     9.9  Meeting

**Drop Duplicates:**


In [28]:
data.drop_duplicates(inplace=True)

In [29]:
data[data.duplicated()]

Empty DataFrame
Columns: [START_DATE*, END_DATE*, CATEGORY*, START*, STOP*, MILES*, PURPOSE*]
Index: []

#### Notice each column label has a `*` at the end?

#### Can we do something to remove this?
- List comprehension


In [30]:
data.columns = [x[:-1] for x in data.columns]

In [31]:
data

START_DATE          END_DATE  CATEGORY             START  \
0       1/1/2016 21:11    1/1/2016 21:17  Business       Fort Pierce   
1        1/2/2016 1:25     1/2/2016 1:37  Business       Fort Pierce   
2       1/2/2016 20:25    1/2/2016 20:38  Business       Fort Pierce   
3       1/5/2016 17:31    1/5/2016 17:45  Business       Fort Pierce   
4       1/6/2016 14:42    1/6/2016 15:49  Business       Fort Pierce   
...                ...               ...       ...               ...   
1151  12/31/2016 13:24  12/31/2016 13:42  Business           Kar?chi   
1152  12/31/2016 15:03  12/31/2016 15:38  Business  Unknown Location   
1153  12/31/2016 21:32  12/31/2016 21:50  Business        Katunayake   
1154  12/31/2016 22:08  12/31/2016 23:51  Business           Gampaha   
1155            Totals               NaN       NaN               NaN   

                  STOP    MILES          PURPOSE  
0          Fort Pierce      5.1   Meal/Entertain  
1          Fort Pierce      5.0              NaN  
2          Fort Pierce      4.8  Errand/Supplies  
3          Fort Pierce      4.7          Meeting  
4      West Palm Beach     63.7   Customer Visit  
...                ...      ...              ...  
1151  Unknown Location      3.9   Temporary Site  
1152  Unknown Location     16.2          Meeting  
1153           Gampaha      6.4   Temporary Site  
1154         Ilukwatta     48.2   Temporary Site  
1155               NaN  12204.7              NaN  

[1155 rows x 7 columns]


### Handling Date-Time data


  - `to_datetime()`
  
  
  - It takes as input:
    - Array/Scalars with values having proper date/time format
    - `dayfirst`: Indicating if the day comes first in the date format used
    - `yearfirst`: Indicates if year comes first in the date format

Refer: https://pandas.pydata.org/docs/reference/api/pandas.to_datetime.html

In [36]:
data = data[data['START_DATE'] != 'Totals'].copy()
data['START_DATE'] = pd.to_datetime(data['START_DATE'])
data['END_DATE'] = pd.to_datetime(data['END_DATE'])

data

START_DATE            END_DATE  CATEGORY             START  \
0    2016-01-01 21:11:00 2016-01-01 21:17:00  Business       Fort Pierce   
1    2016-01-02 01:25:00 2016-01-02 01:37:00  Business       Fort Pierce   
2    2016-01-02 20:25:00 2016-01-02 20:38:00  Business       Fort Pierce   
3    2016-01-05 17:31:00 2016-01-05 17:45:00  Business       Fort Pierce   
4    2016-01-06 14:42:00 2016-01-06 15:49:00  Business       Fort Pierce   
...                  ...                 ...       ...               ...   
1150 2016-12-31 01:07:00 2016-12-31 01:14:00  Business           Kar?chi   
1151 2016-12-31 13:24:00 2016-12-31 13:42:00  Business           Kar?chi   
1152 2016-12-31 15:03:00 2016-12-31 15:38:00  Business  Unknown Location   
1153 2016-12-31 21:32:00 2016-12-31 21:50:00  Business        Katunayake   
1154 2016-12-31 22:08:00 2016-12-31 23:51:00  Business           Gampaha   

                  STOP  MILES          PURPOSE  
0          Fort Pierce    5.1   Meal/Entertain  
1          Fort Pierce    5.0              NaN  
2          Fort Pierce    4.8  Errand/Supplies  
3          Fort Pierce    4.7          Meeting  
4      West Palm Beach   63.7   Customer Visit  
...                ...    ...              ...  
1150           Kar?chi    0.7          Meeting  
1151  Unknown Location    3.9   Temporary Site  
1152  Unknown Location   16.2          Meeting  
1153           Gampaha    6.4   Temporary Site  
1154         Ilukwatta   48.2   Temporary Site  

[1154 rows x 7 columns]

In [37]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1154 entries, 0 to 1154
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   START_DATE  1154 non-null   datetime64[ns]
 1   END_DATE    1154 non-null   datetime64[ns]
 2   CATEGORY    1154 non-null   object        
 3   START       1154 non-null   object        
 4   STOP        1154 non-null   object        
 5   MILES       1154 non-null   float64       
 6   PURPOSE     652 non-null    object        
dtypes: datetime64[ns](2), float64(1), object(4)
memory usage: 72.1+ KB


### Cross Check for rows where start timestamp and end timestamp are same ?

#### Because having exactly zero miles is not  really possible







In [38]:
data[data['START_DATE']==data['END_DATE']]

START_DATE            END_DATE  CATEGORY             START  \
751 2016-09-06 17:49:00 2016-09-06 17:49:00  Business  Unknown Location   
761 2016-09-16 07:08:00 2016-09-16 07:08:00  Business  Unknown Location   
798 2016-10-08 15:03:00 2016-10-08 15:03:00  Business           Karachi   
807 2016-10-13 13:02:00 2016-10-13 13:02:00  Business         Islamabad   

                 STOP  MILES PURPOSE  
751  Unknown Location   69.1     NaN  
761  Unknown Location    1.6     NaN  
798           Karachi    3.6     NaN  
807         Islamabad    0.7     NaN

In [39]:
data.drop(data.index[[751,761,798,807]],inplace=True)
data

START_DATE            END_DATE  CATEGORY             START  \
0    2016-01-01 21:11:00 2016-01-01 21:17:00  Business       Fort Pierce   
1    2016-01-02 01:25:00 2016-01-02 01:37:00  Business       Fort Pierce   
2    2016-01-02 20:25:00 2016-01-02 20:38:00  Business       Fort Pierce   
3    2016-01-05 17:31:00 2016-01-05 17:45:00  Business       Fort Pierce   
4    2016-01-06 14:42:00 2016-01-06 15:49:00  Business       Fort Pierce   
...                  ...                 ...       ...               ...   
1150 2016-12-31 01:07:00 2016-12-31 01:14:00  Business           Kar?chi   
1151 2016-12-31 13:24:00 2016-12-31 13:42:00  Business           Kar?chi   
1152 2016-12-31 15:03:00 2016-12-31 15:38:00  Business  Unknown Location   
1153 2016-12-31 21:32:00 2016-12-31 21:50:00  Business        Katunayake   
1154 2016-12-31 22:08:00 2016-12-31 23:51:00  Business           Gampaha   

                  STOP  MILES          PURPOSE  
0          Fort Pierce    5.1   Meal/Entertain  
1          Fort Pierce    5.0              NaN  
2          Fort Pierce    4.8  Errand/Supplies  
3          Fort Pierce    4.7          Meeting  
4      West Palm Beach   63.7   Customer Visit  
...                ...    ...              ...  
1150           Kar?chi    0.7          Meeting  
1151  Unknown Location    3.9   Temporary Site  
1152  Unknown Location   16.2          Meeting  
1153           Gampaha    6.4   Temporary Site  
1154         Ilukwatta   48.2   Temporary Site  

[1150 rows x 7 columns]

In [40]:
ts = data['END_DATE'][0]
ts

Timestamp('2016-01-01 21:17:00')

#### Now how can we extract the year from this date ?
  - Using the `year` attribute



In [41]:
ts.year

2016

In [42]:
ts.month

1

In [43]:
ts.day

1

#### But what if we want to know the name of the month or the day of the week on that date ?
  - using `month_name()` and `day_name()` methods

In [44]:
ts.month_name()

'January'

In [45]:
ts.day_name()

'Friday'

In [46]:
ts.dayofweek

4

In [48]:
ts.hour

21


- We can similarly extract minutes and seconds


### Let's add a new column having only `year`
- `.dt`

In [49]:
data['END_DATE'].dt

In [51]:
data['END_DATE'].dt.year

0       2016
1       2016
2       2016
3       2016
4       2016
        ... 
1150    2016
1151    2016
1152    2016
1153    2016
1154    2016
Name: END_DATE, Length: 1150, dtype: int32

In [52]:
data['END_DATE'].dt.month_name()

0        January
1        January
2        January
3        January
4        January
          ...   
1150    December
1151    December
1152    December
1153    December
1154    December
Name: END_DATE, Length: 1150, dtype: object

In [53]:
data['year'] = data['END_DATE'].dt.year
data.head()

START_DATE            END_DATE  CATEGORY        START  \
0 2016-01-01 21:11:00 2016-01-01 21:17:00  Business  Fort Pierce   
1 2016-01-02 01:25:00 2016-01-02 01:37:00  Business  Fort Pierce   
2 2016-01-02 20:25:00 2016-01-02 20:38:00  Business  Fort Pierce   
3 2016-01-05 17:31:00 2016-01-05 17:45:00  Business  Fort Pierce   
4 2016-01-06 14:42:00 2016-01-06 15:49:00  Business  Fort Pierce   

              STOP  MILES          PURPOSE  year  
0      Fort Pierce    5.1   Meal/Entertain  2016  
1      Fort Pierce    5.0              NaN  2016  
2      Fort Pierce    4.8  Errand/Supplies  2016  
3      Fort Pierce    4.7          Meeting  2016  
4  West Palm Beach   63.7   Customer Visit  2016

#### Lets try to answer the following questions:

- **What is the shortest journey made?**


- **What is the longest journey made?**


- **What is the average journey made?**


- **How many years of data do we have?**





In [54]:
data.describe()

START_DATE                       END_DATE  \
count                           1150                           1150   
mean   2016-07-16 18:02:01.147825920  2016-07-16 18:25:18.782608896   
min              2016-01-01 21:11:00            2016-01-01 21:17:00   
25%              2016-03-31 13:14:30            2016-03-31 13:48:45   
50%              2016-07-20 18:03:00            2016-07-20 18:08:30   
75%              2016-10-28 08:02:00            2016-10-28 08:21:00   
max              2016-12-31 22:08:00            2016-12-31 23:51:00   
std                              NaN                            NaN   

             MILES    year  
count  1150.000000  1150.0  
mean     10.584609  2016.0  
min       0.500000  2016.0  
25%       2.900000  2016.0  
50%       6.000000  2016.0  
75%      10.400000  2016.0  
max     310.300000  2016.0  
std      21.623241     0.0

#### So We can answer all the questions by looking at it:

- **What is the shortest journey made?** - 0.5 miles


- **What is the longest journey made?** - 310.3 miles


- **What is the average journey made?** - mean value


- **How many years of data do we have?** - Just 1 year - 2016